Get prop manager thing working

DavidFair opened the pull request:
Makes it so property manager will correctly with TypedValues instead of strings everywhere.
This required some nuance as there were also external users of the types so we had to maintain bug compat to not break their stuff

TODO - Run through all tests locally

DavidFair committed:
Use IPropertyManager to manage Alg Props

Explicitly check that props exist when updating

This preserves existing behaviour where it would throw trying to assign,
which existing tests rely on.

DavidFair committed:
Replace new/delete with unique_ptr in PropManagerTest

Since this is the preferred modern version

DavidFair committed:
Add method to get names for properties from PropManager

This avoids callers having to manually parse this (possibly incorrectly)
and is required for some upstream work

DavidFair committed:
Update expected call count for AlgStarted in tests

Since the validation now takes place in execute, rather than when we're
copying the properties. This means from an observer POV we are now
starting the "correct" number of algorithms too.

cailafinn commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.cpp:
```diff
@@ -467,16 +467,16 @@ void RunsPresenter::handleError(const std::string &message) { m_messageHandler->
+  options->setPropertyValue("InputWorkspace", inputWorkspace);
+  options->setPropertyValue("Instrument", instrument);
+  options->setPropertyValue("GetLiveValueAlgorithm", "GetLiveInstrumentValue");
   // Convert the properties to a string to pass to the algorithm
```
This isn't a string anymore

DavidFair commented on qt/scientific_interfaces/ISISReflectometry/GUI/Runs/RunsPresenter.cpp:
```diff
@@ -467,16 +467,16 @@ void RunsPresenter::handleError(const std::string &message) { m_messageHandler->
+  options->setPropertyValue("Instrument", instrument);
+  options->setPropertyValue("GetLiveValueAlgorithm", "GetLiveInstrumentValue");
   // Convert the properties to a string to pass to the algorithm
-  auto const optionsString = convertMapToString(options, ';', false);
```
@gemmaguest Will this cause problems now, since the options is now an actual type instead of a `;` delimited string?

cailafinn commented on qt/widgets/common/inc/MantidQtWidgets/Common/AlgorithmRuntimeProps.h:
```diff
@@ -0,0 +1,29 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2012 ISIS Rutherford Appleton Laboratory UKRI,
```
```suggestion
// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
```

cailafinn commented on qt/widgets/common/inc/MantidQtWidgets/Common/IAlgorithmRuntimeProps.h:
```diff
@@ -0,0 +1,30 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2012 ISIS Rutherford Appleton Laboratory UKRI,
```
```suggestion
// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
```

cailafinn commented on qt/widgets/common/src/AlgorithmRuntimeProps.cpp:
```diff
@@ -0,0 +1,28 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2012 ISIS Rutherford Appleton Laboratory UKRI,
```
```suggestion
// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
```

cailafinn commented on qt/widgets/common/inc/MantidQtWidgets/Common/IConfiguredAlgorithm.h:
```diff
@@ -4,24 +4,22 @@
+  virtual ~IConfiguredAlgorithm() = default;
   virtual Mantid::API::IAlgorithm_sptr algorithm() const = 0;
-  virtual AlgorithmRuntimeProps properties() const = 0;
+  virtual const MantidQt::API::IAlgorithmRuntimeProps &properties() const noexcept = 0;
```
This doesn't get the properties themselves, it gets the AlgRuntimeProps object. This could be renamed so that we don't have to do `Alg->properties().getProperties()` or whatever

cailafinn commented on qt/widgets/common/src/BatchAlgorithmRunner.cpp:
```diff
@@ -201,11 +207,23 @@ bool BatchAlgorithmRunner::executeAlgo(const IConfiguredAlgorithm_sptr &algorith
+    if (invalidProps.size() > 0) {
+      const auto invalidPropsStr = std::accumulate(invalidProps.cbegin(), invalidProps.cend(), invalidProps[0],
+                                                   [](const auto &a, const auto &b) { return a + "," + b; });
+      throw Mantid::Kernel::Exception::NotFoundError("Invalid Properties given:", invalidPropsStr);
```
```suggestion
      throw Mantid::Kernel::Exception::NotFoundError("Invalid Properties given: ", invalidPropsStr);
```

cailafinn commented on qt/widgets/common/src/ConfiguredAlgorithm.cpp:
```diff
@@ -0,0 +1,21 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2012 ISIS Rutherford Appleton Laboratory UKRI,
```
```suggestion
// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
```

cailafinn commented on qt/widgets/common/test/AlgorithmRuntimePropsTest.h:
```diff
@@ -0,0 +1,45 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
```
```suggestion
// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
```

cailafinn commented on qt/widgets/common/test/MockConfiguredAlgorithm.h:
```diff
@@ -7,16 +7,23 @@
+  MockConfiguredAlgorithm(std::unique_ptr<MantidQt::API::IAlgorithmRuntimeProps> runtimeProps)
+      : m_runtimeProps(std::move(runtimeProps)) {
+    ON_CALL(*this, properties).WillByDefault(::testing::ReturnRef(*m_runtimeProps));
+  }
```
This is slightly unconventional mock, so might be worth writing a comment to explain why we're enforcing this.

cailafinn reviewed (changes requested):
Few minor comments and calrifications, but otherwise really good!

DavidFair committed:
Address PR comments on AlgRuntimeProps changes

Various PR comments including:
- Copyright years
- Better method naming
- Some comments in code

DavidFair committed:
Add parser for AlgRuntimeProps to string

This is required to maintain compatibility with the existing
RunPresenter class in reflectometry

DavidFair committed:
Fixup incorrectly ported comparison and tests

Fixes an incorrect comparison, which implicitly relied on string
convertable types and would fail.

Fix a test where I assumed all properties were equal, whereas the
original code simply check the properties exists and we're the same for
the list passed (inclusive)

cailafinn commented on qt/widgets/common/src/BatchAlgorithmRunner.cpp:
```diff
@@ -218,7 +218,7 @@ bool BatchAlgorithmRunner::executeAlgo(const IConfiguredAlgorithm_sptr &algorith
     if (invalidProps.size() > 0) {
       const auto invalidPropsStr = std::accumulate(invalidProps.cbegin(), invalidProps.cend(), invalidProps[0],
                                                    [](const auto &a, const auto &b) { return a + "," + b; });
```
I believe this is going to have the first `invalidProps` value appear twice? The comma could also do with a space after it I think. Also, do we/can we have a test for this?

cailafinn reviewed (changes requested):
Just noticed this, but otherwise looks good!

Commit:
Fix invalid props exception printing first attr twice

This would print the first element name twice incorrectly when there are
n>2 elements, as we did not correctly move the iterator on before
starting std::accumulate

gemmaguest commented on qt/widgets/common/test/BatchAlgorithmRunnerTest.h:
```diff
@@ -227,7 +227,7 @@ class BatchAlgorithmRunnerTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(batchCompleteSpy.count(), 1);
     TS_ASSERT_EQUALS(batchCancelledSpy.count(), 0);
-    TS_ASSERT_EQUALS(algStartSpy.count(), 1);
+    TS_ASSERT_EQUALS(algStartSpy.count(), 2);
```
How come the count on these signals has changed?

cailafinn replied:
Validation no longer happens when setting props, only on starting; so one extra starts but the same number completes
